cache param env canonicalization

aliemjay opened the pull request:
Canonicalize ParamEnv only once and store it. Then whenever we try to canonicalize `ParamEnvAnd<'tcx, T>` we only have to canonicalize `T` and then merge the results.

Prelimiary results show ~3-4% savings in diesel and serde benchmarks.

Best to review commits individually. Some commits have a short description.

Initial implementation had a soundness bug (https://github.com/rust-lang/rust/pull/117749#issuecomment-1840453387) due to cache invalidation:
- When canonicalizing `Ty<'?0>` we first try to resolve region variables in the current InferCtxt which may have a constraint `?0 == 'static`. This means that we register `Ty<'?0> => Canonical<Ty<'static>>` in the cache, which is obviously incorrect in another inference context.
- This is fixed by not doing region resolution when canonicalizing the query *input* (vs. response), which is the only place where ParamEnv is used, and then in a later commit we *statically* guard against any form of inference variable resolution of the cached canonical ParamEnv's.

r? @ghost

aliemjay commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit b93a98e0fca57c85bdf1ccbb96efb85d4b8d7d42 with merge 557c12e39628232629cbe904f927f16144010a47...
<!-- homu: {"type":"TryBuildStarted","head_sha":"b93a98e0fca57c85bdf1ccbb96efb85d4b8d7d42","merge_sha":"557c12e39628232629cbe904f927f16144010a47"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/6811074483/job/18523370049)
Build commit: 557c12e39628232629cbe904f927f16144010a47 (`557c12e39628232629cbe904f927f16144010a47`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/6811074483/job/18523370049"},"merge_sha":"557c12e39628232629cbe904f927f16144010a47"} -->

rust-timer commented:
Queued 557c12e39628232629cbe904f927f16144010a47 with parent 287ae4db75228ef8260a112bcc65245f05e18340, future [comparison URL](https://perf.rust-lang.org/compare.html?start=287ae4db75228ef8260a112bcc65245f05e18340&end=557c12e39628232629cbe904f927f16144010a47).
There is currently 1 preceding artifact in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~2.5 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (557c12e39628232629cbe904f927f16144010a47): [comparison URL](https://perf.rust-lang.org/compare.html?start=287ae4db75228ef8260a112bcc65245f05e18340&end=557c12e39628232629cbe904f927f16144010a47&stat=instructions:u).

### Overall result: ✅ improvements - no action needed


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

@bors rollup=never
@rustbot label: -S-waiting-on-perf -perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=287ae4db75228ef8260a112bcc65245f05e18340&end=557c12e39628232629cbe904f927f16144010a47&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.3%  | [0.3%, 0.3%]   | 1     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -1.9% | [-3.3%, -0.4%] | 17    |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -1.7% | [-3.3%, 0.3%]  | 18    |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=287ae4db75228ef8260a112bcc65245f05e18340&end=557c12e39628232629cbe904f927f16144010a47&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.7%  | [0.6%, 0.8%]   | 2     |
| Regressions ❌ <br /> (secondary)  | 4.7%  | [4.3%, 5.2%]   | 2     |
| Improvements ✅ <br /> (primary)   | -1.2% | [-4.2%, -0.4%] | 11    |
| Improvements ✅ <br /> (secondary) | -2.0% | [-2.0%, -2.0%] | 1     |
| All ❌✅ (primary)                 | -1.0% | [-4.2%, 0.8%]  | 13    |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=287ae4db75228ef8260a112bcc65245f05e18340&end=557c12e39628232629cbe904f927f16144010a47&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.5%  | [0.5%, 0.5%]   | 1     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.9% | [-1.3%, -0.5%] | 6     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -0.7% | [-1.3%, 0.5%]  | 7     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=287ae4db75228ef8260a112bcc65245f05e18340&end=557c12e39628232629cbe904f927f16144010a47&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 661.067s -> 661.537s (0.07%)
**Artifact size**: 308.78 MiB -> 308.75 MiB (-0.01%)

aliemjay committed:
global param_env canonicalization cache

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit bcbfbe62ff1ca24302e585679473c43f59054717 with merge 3d74566bec1c4a843859581f252317eb7b1bacbf...
<!-- homu: {"type":"TryBuildStarted","head_sha":"bcbfbe62ff1ca24302e585679473c43f59054717","merge_sha":"3d74566bec1c4a843859581f252317eb7b1bacbf"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/7092350207/job/19303804736)
<!-- homu: {"type":"TryBuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/7092350207/job/19303804736","builder_name":"checks-actions"} -->

rust-log-analyzer commented:
The job **`dist-x86_64-linux`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/7092350207/job/19303457387) [(plain)](https://github.com/rust-lang-ci/rust/commit/3d74566bec1c4a843859581f252317eb7b1bacbf/checks/19303457387/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
                               at /rustc/3d74566bec1c4a843859581f252317eb7b1bacbf/library/std/src/sys_common/backtrace.rs:34:9
   9:     0x7fb368eacf57 - std::panicking::default_hook::{{closure}}::ha6969053640e854b
  10:     0x7fb368eaccbf - std::panicking::default_hook::hb8c2992d65e2a1fd
                               at /rustc/3d74566bec1c4a843859581f252317eb7b1bacbf/library/std/src/panicking.rs:292:9
  11:     0x7fb366619525 - std[3cc7befc00c2e885]::panicking::update_hook::<alloc[14f9cae0e30d7e5f]::boxed::Box<rustc_driver_impl[a104305ea8c6183c]::install_ice_hook::{closure#0}>>::{closure#0}
  12:     0x7fb368ead698 - <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call::ha25fbd5d0c172899
  13:     0x7fb368ead698 - std::panicking::rust_panic_with_hook::h7e4df4fcf0c4002f
                               at /rustc/3d74566bec1c4a843859581f252317eb7b1bacbf/library/std/src/panicking.rs:783:13
  14:     0x7fb368ead3ee - std::panicking::begin_panic_handler::{{closure}}::hfc2e5d86d9a119a7
                               at /rustc/3d74566bec1c4a843859581f252317eb7b1bacbf/library/std/src/panicking.rs:657:13
---
  17:     0x7fb368ef9be5 - core::panicking::panic_fmt::h33e87dd7bfffedd5
                               at /rustc/3d74566bec1c4a843859581f252317eb7b1bacbf/library/core/src/panicking.rs:72:14
  18:     0x7fb368ef9e12 - core::panicking::panic_bounds_check::hf8181e21673d41b8
                               at /rustc/3d74566bec1c4a843859581f252317eb7b1bacbf/library/core/src/panicking.rs:205:5
  19:     0x7fb367e1fd47 - <ena[d7f1b039fa591958]::unify::UnificationTable<ena[d7f1b039fa591958]::unify::backing_vec::InPlace<rustc_middle[99d1f70202f125a9]::infer::unify_key::RegionVidKey, &mut alloc[14f9cae0e30d7e5f]::vec::Vec<ena[d7f1b039fa591958]::unify::VarValue<rustc_middle[99d1f70202f125a9]::infer::unify_key::RegionVidKey>>, &mut rustc_infer[5e6a8f20693c1ec]::infer::undo_log::InferCtxtUndoLogs>>>::uninlined_get_root_key
  20:     0x7fb367e50f6c - <rustc_infer[5e6a8f20693c1ec]::infer::equate::Equate as rustc_middle[99d1f70202f125a9]::ty::relate::TypeRelation>::regions
  21:     0x7fb367bdb71d - <rustc_infer[5e6a8f20693c1ec]::infer::at::At>::eq::<rustc_middle[99d1f70202f125a9]::ty::sty::Region>
  22:     0x7fb367bc24ce - <rustc_infer[5e6a8f20693c1ec]::infer::InferCtxt>::instantiate_query_response_and_region_obligations::<rustc_middle[99d1f70202f125a9]::traits::query::NormalizationResult>
  23:     0x7fb367d6e9fd - <rustc_trait_selection[4e49e628b386f5c5]::traits::query::normalize::QueryNormalizer as rustc_type_ir[89060da0ef9e9ec0]::fold::FallibleTypeFolder<rustc_middle[99d1f70202f125a9]::ty::context::TyCtxt>>::try_fold_ty
  24:     0x7fb367755e9a - <rustc_middle[99d1f70202f125a9]::ty::sty::FnSig as rustc_type_ir[89060da0ef9e9ec0]::fold::TypeFoldable<rustc_middle[99d1f70202f125a9]::ty::context::TyCtxt>>::try_fold_with::<rustc_trait_selection[4e49e628b386f5c5]::traits::query::normalize::QueryNormalizer>
  25:     0x7fb367772df1 - rustc_traits[cf6a4930308f040e]::type_op::type_op_normalize_fn_sig
  26:     0x7fb367a8aa89 - rustc_query_impl[e2bf5c9a174b90c7]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[e2bf5c9a174b90c7]::query_impl::type_op_normalize_fn_sig::dynamic_query::{closure#2}::{closure#0}, rustc_middle[99d1f70202f125a9]::query::erase::Erased<[u8; 8usize]>>
  27:     0x7fb367a61fc0 - <rustc_query_impl[e2bf5c9a174b90c7]::query_impl::type_op_normalize_fn_sig::dynamic_query::{closure#2} as core[8a36ddf6fdea383d]::ops::function::FnOnce<(rustc_middle[99d1f70202f125a9]::ty::context::TyCtxt, rustc_type_ir[89060da0ef9e9ec0]::canonical::Canonical<rustc_middle[99d1f70202f125a9]::ty::context::TyCtxt, rustc_middle[99d1f70202f125a9]::ty::ParamEnvAnd<rustc_middle[99d1f70202f125a9]::traits::query::type_op::Normalize<rustc_middle[99d1f70202f125a9]::ty::sty::FnSig>>>)>>::call_once
  28:     0x7fb36786f2f5 - rustc_query_system[64295cc4866c87b6]::query::plumbing::try_execute_query::<rustc_query_impl[e2bf5c9a174b90c7]::DynamicConfig<rustc_query_system[64295cc4866c87b6]::query::caches::DefaultCache<rustc_type_ir[89060da0ef9e9ec0]::canonical::Canonical<rustc_middle[99d1f70202f125a9]::ty::context::TyCtxt, rustc_middle[99d1f70202f125a9]::ty::ParamEnvAnd<rustc_middle[99d1f70202f125a9]::traits::query::type_op::Normalize<rustc_middle[99d1f70202f125a9]::ty::sty::FnSig>>>, rustc_middle[99d1f70202f125a9]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[e2bf5c9a174b90c7]::plumbing::QueryCtxt, false>
  29:     0x7fb367ae7fc9 - rustc_query_impl[e2bf5c9a174b90c7]::query_impl::type_op_normalize_fn_sig::get_query_non_incr::__rust_end_short_backtrace
  30:     0x7fb367d35030 - <rustc_middle[99d1f70202f125a9]::ty::sty::FnSig as rustc_trait_selection[4e49e628b386f5c5]::traits::query::type_op::normalize::Normalizable>::type_op_method
  31:     0x7fb36762ffff - <rustc_middle[99d1f70202f125a9]::ty::ParamEnvAnd<rustc_middle[99d1f70202f125a9]::traits::query::type_op::Normalize<rustc_middle[99d1f70202f125a9]::ty::sty::FnSig>> as rustc_trait_selection[4e49e628b386f5c5]::traits::query::type_op::TypeOp>::fully_perform
  32:     0x7fb3677398b4 - <rustc_borrowck[82aecdca7cf9e523]::type_check::TypeChecker>::typeck_mir
  33:     0x7fb3676f6f26 - rustc_borrowck[82aecdca7cf9e523]::type_check::type_check
  34:     0x7fb3676cadc8 - rustc_borrowck[82aecdca7cf9e523]::nll::compute_regions
  35:     0x7fb36773e916 - rustc_borrowck[82aecdca7cf9e523]::do_mir_borrowck
  36:     0x7fb367709178 - rustc_borrowck[82aecdca7cf9e523]::mir_borrowck
  37:     0x7fb367a86bf7 - rustc_query_impl[e2bf5c9a174b90c7]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[e2bf5c9a174b90c7]::query_impl::mir_borrowck::dynamic_query::{closure#2}::{closure#0}, rustc_middle[99d1f70202f125a9]::query::erase::Erased<[u8; 8usize]>>
  38:     0x7fb367913800 - rustc_query_system[64295cc4866c87b6]::query::plumbing::try_execute_query::<rustc_query_impl[e2bf5c9a174b90c7]::DynamicConfig<rustc_query_system[64295cc4866c87b6]::query::caches::VecCache<rustc_span[9688fd9577add11f]::def_id::LocalDefId, rustc_middle[99d1f70202f125a9]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[e2bf5c9a174b90c7]::plumbing::QueryCtxt, false>
  39:     0x7fb367aae02a - rustc_query_impl[e2bf5c9a174b90c7]::query_impl::mir_borrowck::get_query_non_incr::__rust_end_short_backtrace
  40:     0x7fb366780bc7 - <rustc_middle[99d1f70202f125a9]::hir::map::Map>::par_body_owners::<rustc_interface[144b0912e9fa0e54]::passes::analysis::{closure#1}::{closure#0}>::{closure#0}
  42:     0x7fb367a8b4e7 - rustc_query_impl[e2bf5c9a174b90c7]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[e2bf5c9a174b90c7]::query_impl::analysis::dynamic_query::{closure#2}::{closure#0}, rustc_middle[99d1f70202f125a9]::query::erase::Erased<[u8; 1usize]>>
  43:     0x7fb367860cbf - rustc_query_system[64295cc4866c87b6]::query::plumbing::try_execute_query::<rustc_query_impl[e2bf5c9a174b90c7]::DynamicConfig<rustc_query_system[64295cc4866c87b6]::query::caches::SingleCache<rustc_middle[99d1f70202f125a9]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[e2bf5c9a174b90c7]::plumbing::QueryCtxt, false>
  44:     0x7fb367a94ae7 - rustc_query_impl[e2bf5c9a174b90c7]::query_impl::analysis::get_query_non_incr::__rust_end_short_backtrace
  45:     0x7fb3665c4736 - <rustc_interface[144b0912e9fa0e54]::queries::QueryResult<&rustc_middle[99d1f70202f125a9]::ty::context::GlobalCtxt>>::enter::<core[8a36ddf6fdea383d]::result::Result<(), rustc_span[9688fd9577add11f]::ErrorGuaranteed>, rustc_driver_impl[a104305ea8c6183c]::run_compiler::{closure#0}::{closure#0}::{closure#3}>
  45:     0x7fb3665c4736 - <rustc_interface[144b0912e9fa0e54]::queries::QueryResult<&rustc_middle[99d1f70202f125a9]::ty::context::GlobalCtxt>>::enter::<core[8a36ddf6fdea383d]::result::Result<(), rustc_span[9688fd9577add11f]::ErrorGuaranteed>, rustc_driver_impl[a104305ea8c6183c]::run_compiler::{closure#0}::{closure#0}::{closure#3}>
  46:     0x7fb366617f2c - rustc_interface[144b0912e9fa0e54]::interface::run_compiler::<core[8a36ddf6fdea383d]::result::Result<(), rustc_span[9688fd9577add11f]::ErrorGuaranteed>, rustc_driver_impl[a104305ea8c6183c]::run_compiler::{closure#0}>::{closure#0}
  47:     0x7fb366602e13 - std[3cc7befc00c2e885]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[144b0912e9fa0e54]::util::run_in_thread_with_globals<rustc_interface[144b0912e9fa0e54]::util::run_in_thread_pool_with_globals<rustc_interface[144b0912e9fa0e54]::interface::run_compiler<core[8a36ddf6fdea383d]::result::Result<(), rustc_span[9688fd9577add11f]::ErrorGuaranteed>, rustc_driver_impl[a104305ea8c6183c]::run_compiler::{closure#0}>::{closure#0}, core[8a36ddf6fdea383d]::result::Result<(), rustc_span[9688fd9577add11f]::ErrorGuaranteed>>::{closure#0}, core[8a36ddf6fdea383d]::result::Result<(), rustc_span[9688fd9577add11f]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[8a36ddf6fdea383d]::result::Result<(), rustc_span[9688fd9577add11f]::ErrorGuaranteed>>
  48:     0x7fb36661c0de - <<std[3cc7befc00c2e885]::thread::Builder>::spawn_unchecked_<rustc_interface[144b0912e9fa0e54]::util::run_in_thread_with_globals<rustc_interface[144b0912e9fa0e54]::util::run_in_thread_pool_with_globals<rustc_interface[144b0912e9fa0e54]::interface::run_compiler<core[8a36ddf6fdea383d]::result::Result<(), rustc_span[9688fd9577add11f]::ErrorGuaranteed>, rustc_driver_impl[a104305ea8c6183c]::run_compiler::{closure#0}>::{closure#0}, core[8a36ddf6fdea383d]::result::Result<(), rustc_span[9688fd9577add11f]::ErrorGuaranteed>>::{closure#0}, core[8a36ddf6fdea383d]::result::Result<(), rustc_span[9688fd9577add11f]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[8a36ddf6fdea383d]::result::Result<(), rustc_span[9688fd9577add11f]::ErrorGuaranteed>>::{closure#1} as core[8a36ddf6fdea383d]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  49:     0x7fb368eb7525 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::hd1bdc4496162a15a
                               at /rustc/3d74566bec1c4a843859581f252317eb7b1bacbf/library/alloc/src/boxed.rs:2007:9
  50:     0x7fb368eb7525 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h8b5c739378b82ea9
  51:     0x7fb368eb7525 - std::sys::unix::thread::Thread::new::thread_start::h19a0a8de0640e5ea
                               at /rustc/3d74566bec1c4a843859581f252317eb7b1bacbf/library/std/src/sys/unix/thread.rs:108:17
  52:     0x7fb3651c1ea5 - start_thread
  53:     0x7fb364eeab0d - clone
  53:     0x7fb364eeab0d - clone
  54:                0x0 - <unknown>

error: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please attach the file at `/cargo/registry/src/index.crates.io-6f17d22bba15001f/zerovec-0.10.0/rustc-ice-2023-12-04T20_57_11-46268.txt` to your bug report

note: compiler flags: --crate-type lib -C embed-bitcode=no -C debug-assertions=off -Z unstable-options -C linker=clang -Z binary-dep-depinfo -Z unstable-options -C linker=clang -C link-args=-fuse-ld=lld -Z force-unstable-if-unmarked
note: some of the compiler flags provided by cargo are hidden

query stack during panic:
query stack during panic:
#0 [type_op_normalize_fn_sig] normalizing `extern "RustCall" fn(F/#2, (&ReBound(DebruijnIndex(0), BoundRegion { var: 1, kind: BrAnon }) mut varzerovec::vec::VarZeroVec<ReBound(DebruijnIndex(0), BoundRegion { var: 2, kind: BrAnon }), T/#1, varzerovec::components::Index16>,)) -> Alias(Projection, AliasTy { args: [F/#2, (&ReBound(DebruijnIndex(0), BoundRegion { var: 1, kind: BrAnon }) mut varzerovec::vec::VarZeroVec<ReBound(DebruijnIndex(0), BoundRegion { var: 2, kind: BrAnon }), T/#1, varzerovec::components::Index16>,)], def_id: DefId(1:3085 ~ core[fa13]::ops::function::FnOnce::Output) })`
#1 [mir_borrowck] borrow-checking `yoke_impls::<impl at /rust/deps/zerovec-0.10.0/src/yoke_impls.rs:51:1: 51:86>::transform_mut`
end of query stack
[RUSTC-TIMING] zerovec test:false 0.698
error: could not compile `zerovec` (lib)
warning: build failed, waiting for other jobs to finish...
---
                               at /rustc/3d74566bec1c4a843859581f252317eb7b1bacbf/library/std/src/sys_common/backtrace.rs:34:9
   9:     0x7f4cbf2c3f57 - std::panicking::default_hook::{{closure}}::ha6969053640e854b
  10:     0x7f4cbf2c3cbf - std::panicking::default_hook::hb8c2992d65e2a1fd
                               at /rustc/3d74566bec1c4a843859581f252317eb7b1bacbf/library/std/src/panicking.rs:292:9
  11:     0x7f4cbca30525 - std[3cc7befc00c2e885]::panicking::update_hook::<alloc[14f9cae0e30d7e5f]::boxed::Box<rustc_driver_impl[a104305ea8c6183c]::install_ice_hook::{closure#0}>>::{closure#0}
  12:     0x7f4cbf2c4698 - <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call::ha25fbd5d0c172899
  13:     0x7f4cbf2c4698 - std::panicking::rust_panic_with_hook::h7e4df4fcf0c4002f
                               at /rustc/3d74566bec1c4a843859581f252317eb7b1bacbf/library/std/src/panicking.rs:783:13
  14:     0x7f4cbf2c43ee - std::panicking::begin_panic_handler::{{closure}}::hfc2e5d86d9a119a7
                               at /rustc/3d74566bec1c4a843859581f252317eb7b1bacbf/library/std/src/panicking.rs:657:13
---
  17:     0x7f4cbf310be5 - core::panicking::panic_fmt::h33e87dd7bfffedd5
                               at /rustc/3d74566bec1c4a843859581f252317eb7b1bacbf/library/core/src/panicking.rs:72:14
  18:     0x7f4cbf310e12 - core::panicking::panic_bounds_check::hf8181e21673d41b8
                               at /rustc/3d74566bec1c4a843859581f252317eb7b1bacbf/library/core/src/panicking.rs:205:5
  19:     0x7f4cbe236d47 - <ena[d7f1b039fa591958]::unify::UnificationTable<ena[d7f1b039fa591958]::unify::backing_vec::InPlace<rustc_middle[99d1f70202f125a9]::infer::unify_key::RegionVidKey, &mut alloc[14f9cae0e30d7e5f]::vec::Vec<ena[d7f1b039fa591958]::unify::VarValue<rustc_middle[99d1f70202f125a9]::infer::unify_key::RegionVidKey>>, &mut rustc_infer[5e6a8f20693c1ec]::infer::undo_log::InferCtxtUndoLogs>>>::uninlined_get_root_key
  20:     0x7f4cbe267f6c - <rustc_infer[5e6a8f20693c1ec]::infer::equate::Equate as rustc_middle[99d1f70202f125a9]::ty::relate::TypeRelation>::regions
  21:     0x7f4cbdff271d - <rustc_infer[5e6a8f20693c1ec]::infer::at::At>::eq::<rustc_middle[99d1f70202f125a9]::ty::sty::Region>
  22:     0x7f4cbdfd94ce - <rustc_infer[5e6a8f20693c1ec]::infer::InferCtxt>::instantiate_query_response_and_region_obligations::<rustc_middle[99d1f70202f125a9]::traits::query::NormalizationResult>
  23:     0x7f4cbe1859fd - <rustc_trait_selection[4e49e628b386f5c5]::traits::query::normalize::QueryNormalizer as rustc_type_ir[89060da0ef9e9ec0]::fold::FallibleTypeFolder<rustc_middle[99d1f70202f125a9]::ty::context::TyCtxt>>::try_fold_ty
  24:     0x7f4cbdb6ce9a - <rustc_middle[99d1f70202f125a9]::ty::sty::FnSig as rustc_type_ir[89060da0ef9e9ec0]::fold::TypeFoldable<rustc_middle[99d1f70202f125a9]::ty::context::TyCtxt>>::try_fold_with::<rustc_trait_selection[4e49e628b386f5c5]::traits::query::normalize::QueryNormalizer>
  25:     0x7f4cbdb89df1 - rustc_traits[cf6a4930308f040e]::type_op::type_op_normalize_fn_sig
  26:     0x7f4cbdea1a89 - rustc_query_impl[e2bf5c9a174b90c7]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[e2bf5c9a174b90c7]::query_impl::type_op_normalize_fn_sig::dynamic_query::{closure#2}::{closure#0}, rustc_middle[99d1f70202f125a9]::query::erase::Erased<[u8; 8usize]>>
  27:     0x7f4cbde78fc0 - <rustc_query_impl[e2bf5c9a174b90c7]::query_impl::type_op_normalize_fn_sig::dynamic_query::{closure#2} as core[8a36ddf6fdea383d]::ops::function::FnOnce<(rustc_middle[99d1f70202f125a9]::ty::context::TyCtxt, rustc_type_ir[89060da0ef9e9ec0]::canonical::Canonical<rustc_middle[99d1f70202f125a9]::ty::context::TyCtxt, rustc_middle[99d1f70202f125a9]::ty::ParamEnvAnd<rustc_middle[99d1f70202f125a9]::traits::query::type_op::Normalize<rustc_middle[99d1f70202f125a9]::ty::sty::FnSig>>>)>>::call_once
  28:     0x7f4cbdc862f5 - rustc_query_system[64295cc4866c87b6]::query::plumbing::try_execute_query::<rustc_query_impl[e2bf5c9a174b90c7]::DynamicConfig<rustc_query_system[64295cc4866c87b6]::query::caches::DefaultCache<rustc_type_ir[89060da0ef9e9ec0]::canonical::Canonical<rustc_middle[99d1f70202f125a9]::ty::context::TyCtxt, rustc_middle[99d1f70202f125a9]::ty::ParamEnvAnd<rustc_middle[99d1f70202f125a9]::traits::query::type_op::Normalize<rustc_middle[99d1f70202f125a9]::ty::sty::FnSig>>>, rustc_middle[99d1f70202f125a9]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[e2bf5c9a174b90c7]::plumbing::QueryCtxt, false>
  29:     0x7f4cbdefefc9 - rustc_query_impl[e2bf5c9a174b90c7]::query_impl::type_op_normalize_fn_sig::get_query_non_incr::__rust_end_short_backtrace
  30:     0x7f4cbe14c030 - <rustc_middle[99d1f70202f125a9]::ty::sty::FnSig as rustc_trait_selection[4e49e628b386f5c5]::traits::query::type_op::normalize::Normalizable>::type_op_method
  31:     0x7f4cbda46fff - <rustc_middle[99d1f70202f125a9]::ty::ParamEnvAnd<rustc_middle[99d1f70202f125a9]::traits::query::type_op::Normalize<rustc_middle[99d1f70202f125a9]::ty::sty::FnSig>> as rustc_trait_selection[4e49e628b386f5c5]::traits::query::type_op::TypeOp>::fully_perform
  32:     0x7f4cbdb508b4 - <rustc_borrowck[82aecdca7cf9e523]::type_check::TypeChecker>::typeck_mir
  33:     0x7f4cbdb0df26 - rustc_borrowck[82aecdca7cf9e523]::type_check::type_check
  34:     0x7f4cbdae1dc8 - rustc_borrowck[82aecdca7cf9e523]::nll::compute_regions
  35:     0x7f4cbdb55916 - rustc_borrowck[82aecdca7cf9e523]::do_mir_borrowck
  36:     0x7f4cbdb20178 - rustc_borrowck[82aecdca7cf9e523]::mir_borrowck
  37:     0x7f4cbde9dbf7 - rustc_query_impl[e2bf5c9a174b90c7]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[e2bf5c9a174b90c7]::query_impl::mir_borrowck::dynamic_query::{closure#2}::{closure#0}, rustc_middle[99d1f70202f125a9]::query::erase::Erased<[u8; 8usize]>>
  38:     0x7f4cbdd2a800 - rustc_query_system[64295cc4866c87b6]::query::plumbing::try_execute_query::<rustc_query_impl[e2bf5c9a174b90c7]::DynamicConfig<rustc_query_system[64295cc4866c87b6]::query::caches::VecCache<rustc_span[9688fd9577add11f]::def_id::LocalDefId, rustc_middle[99d1f70202f125a9]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[e2bf5c9a174b90c7]::plumbing::QueryCtxt, false>
  39:     0x7f4cbdec502a - rustc_query_impl[e2bf5c9a174b90c7]::query_impl::mir_borrowck::get_query_non_incr::__rust_end_short_backtrace
  40:     0x7f4cbcb97bc7 - <rustc_middle[99d1f70202f125a9]::hir::map::Map>::par_body_owners::<rustc_interface[144b0912e9fa0e54]::passes::analysis::{closure#1}::{closure#0}>::{closure#0}
  42:     0x7f4cbdea24e7 - rustc_query_impl[e2bf5c9a174b90c7]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[e2bf5c9a174b90c7]::query_impl::analysis::dynamic_query::{closure#2}::{closure#0}, rustc_middle[99d1f70202f125a9]::query::erase::Erased<[u8; 1usize]>>
  43:     0x7f4cbdc77cbf - rustc_query_system[64295cc4866c87b6]::query::plumbing::try_execute_query::<rustc_query_impl[e2bf5c9a174b90c7]::DynamicConfig<rustc_query_system[64295cc4866c87b6]::query::caches::SingleCache<rustc_middle[99d1f70202f125a9]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[e2bf5c9a174b90c7]::plumbing::QueryCtxt, false>
  44:     0x7f4cbdeabae7 - rustc_query_impl[e2bf5c9a174b90c7]::query_impl::analysis::get_query_non_incr::__rust_end_short_backtrace
  45:     0x7f4cbc9db736 - <rustc_interface[144b0912e9fa0e54]::queries::QueryResult<&rustc_middle[99d1f70202f125a9]::ty::context::GlobalCtxt>>::enter::<core[8a36ddf6fdea383d]::result::Result<(), rustc_span[9688fd9577add11f]::ErrorGuaranteed>, rustc_driver_impl[a104305ea8c6183c]::run_compiler::{closure#0}::{closure#0}::{closure#3}>
  45:     0x7f4cbc9db736 - <rustc_interface[144b0912e9fa0e54]::queries::QueryResult<&rustc_middle[99d1f70202f125a9]::ty::context::GlobalCtxt>>::enter::<core[8a36ddf6fdea383d]::result::Result<(), rustc_span[9688fd9577add11f]::ErrorGuaranteed>, rustc_driver_impl[a104305ea8c6183c]::run_compiler::{closure#0}::{closure#0}::{closure#3}>
  46:     0x7f4cbca2ef2c - rustc_interface[144b0912e9fa0e54]::interface::run_compiler::<core[8a36ddf6fdea383d]::result::Result<(), rustc_span[9688fd9577add11f]::ErrorGuaranteed>, rustc_driver_impl[a104305ea8c6183c]::run_compiler::{closure#0}>::{closure#0}
  47:     0x7f4cbca19e13 - std[3cc7befc00c2e885]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[144b0912e9fa0e54]::util::run_in_thread_with_globals<rustc_interface[144b0912e9fa0e54]::util::run_in_thread_pool_with_globals<rustc_interface[144b0912e9fa0e54]::interface::run_compiler<core[8a36ddf6fdea383d]::result::Result<(), rustc_span[9688fd9577add11f]::ErrorGuaranteed>, rustc_driver_impl[a104305ea8c6183c]::run_compiler::{closure#0}>::{closure#0}, core[8a36ddf6fdea383d]::result::Result<(), rustc_span[9688fd9577add11f]::ErrorGuaranteed>>::{closure#0}, core[8a36ddf6fdea383d]::result::Result<(), rustc_span[9688fd9577add11f]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[8a36ddf6fdea383d]::result::Result<(), rustc_span[9688fd9577add11f]::ErrorGuaranteed>>
  48:     0x7f4cbca330de - <<std[3cc7befc00c2e885]::thread::Builder>::spawn_unchecked_<rustc_interface[144b0912e9fa0e54]::util::run_in_thread_with_globals<rustc_interface[144b0912e9fa0e54]::util::run_in_thread_pool_with_globals<rustc_interface[144b0912e9fa0e54]::interface::run_compiler<core[8a36ddf6fdea383d]::result::Result<(), rustc_span[9688fd9577add11f]::ErrorGuaranteed>, rustc_driver_impl[a104305ea8c6183c]::run_compiler::{closure#0}>::{closure#0}, core[8a36ddf6fdea383d]::result::Result<(), rustc_span[9688fd9577add11f]::ErrorGuaranteed>>::{closure#0}, core[8a36ddf6fdea383d]::result::Result<(), rustc_span[9688fd9577add11f]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[8a36ddf6fdea383d]::result::Result<(), rustc_span[9688fd9577add11f]::ErrorGuaranteed>>::{closure#1} as core[8a36ddf6fdea383d]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  49:     0x7f4cbf2ce525 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::hd1bdc4496162a15a
                               at /rustc/3d74566bec1c4a843859581f252317eb7b1bacbf/library/alloc/src/boxed.rs:2007:9
  50:     0x7f4cbf2ce525 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h8b5c739378b82ea9
  51:     0x7f4cbf2ce525 - std::sys::unix::thread::Thread::new::thread_start::h19a0a8de0640e5ea
                               at /rustc/3d74566bec1c4a843859581f252317eb7b1bacbf/library/std/src/sys/unix/thread.rs:108:17
  52:     0x7f4cbb5d8ea5 - start_thread
  53:     0x7f4cbb301b0d - clone
  53:     0x7f4cbb301b0d - clone
  54:                0x0 - <unknown>

error: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please attach the file at `/cargo/registry/src/index.crates.io-6f17d22bba15001f/zerovec-0.10.0/rustc-ice-2023-12-04T20_57_11-46291.txt` to your bug report

note: compiler flags: --crate-type lib -C opt-level=3 -C embed-bitcode=no -C codegen-units=1 -Z unstable-options -C linker=clang -C symbol-mangling-version=v0 -Z unstable-options -Z macro-backtrace -C link-args=-Wl,-z,origin -C link-args=-Wl,-rpath,$ORIGIN/../lib -C link-args=-fuse-ld=lld -C split-debuginfo=off -Z unstable-options -C prefer-dynamic -C link-args=-Wl,--icf=all -C profile-generate=/tmp/tmp-multistage/opt-artifacts/rustc-pgo -C llvm-args=-vp-counters-per-site=4 -C llvm-args=-static-func-strip-dirname-prefix=2 -Z dylib-lto -C lto=thin -C embed-bitcode=yes -Z binary-dep-depinfo -Z tls-model=initial-exec -Z force-unstable-if-unmarked
note: some of the compiler flags provided by cargo are hidden

query stack during panic:
query stack during panic:
#0 [type_op_normalize_fn_sig] normalizing `extern "RustCall" fn(F/#2, (&ReBound(DebruijnIndex(0), BoundRegion { var: 1, kind: BrAnon }) mut varzerovec::vec::VarZeroVec<ReBound(DebruijnIndex(0), BoundRegion { var: 2, kind: BrAnon }), T/#1, varzerovec::components::Index16>,)) -> Alias(Projection, AliasTy { args: [F/#2, (&ReBound(DebruijnIndex(0), BoundRegion { var: 1, kind: BrAnon }) mut varzerovec::vec::VarZeroVec<ReBound(DebruijnIndex(0), BoundRegion { var: 2, kind: BrAnon }), T/#1, varzerovec::components::Index16>,)], def_id: DefId(1:3085 ~ core[fa13]::ops::function::FnOnce::Output) })`
#1 [mir_borrowck] borrow-checking `yoke_impls::<impl at /rust/deps/zerovec-0.10.0/src/yoke_impls.rs:51:1: 51:86>::transform_mut`
end of query stack
[RUSTC-TIMING] zerovec test:false 0.912
error: could not compile `zerovec` (lib)
[RUSTC-TIMING] regex_syntax test:false 10.607
---
Caused by:
    Command RUST_BACKTRACE=full python3 /checkout/x.py build --target x86_64-unknown-linux-gnu --host x86_64-unknown-linux-gnu --stage 2 library/std --rust-profile-generate /tmp/tmp-multistage/opt-artifacts/rustc-pgo --set llvm.thin-lto=false --set llvm.link-shared=true [at /checkout/obj] has failed with exit code Some(1)

Stack backtrace:
   0: <opt_dist::exec::CmdBuilder>::run
             at /rustc/3d74566bec1c4a843859581f252317eb7b1bacbf/src/tools/opt-dist/src/exec.rs:78:17
   1: <opt_dist::exec::Bootstrap>::run
             at /rustc/3d74566bec1c4a843859581f252317eb7b1bacbf/src/tools/opt-dist/src/exec.rs:179:9
   2: opt_dist::execute_pipeline::{closure#1}::{closure#0}
             at /rustc/3d74566bec1c4a843859581f252317eb7b1bacbf/src/tools/opt-dist/src/main.rs:210:13
      <opt_dist::timer::TimerSection>::section::<opt_dist::execute_pipeline::{closure#1}::{closure#0}, ()>
             at /rustc/3d74566bec1c4a843859581f252317eb7b1bacbf/src/tools/opt-dist/src/timer.rs:111:22
      opt_dist::execute_pipeline::{closure#1}
             at /rustc/3d74566bec1c4a843859581f252317eb7b1bacbf/src/tools/opt-dist/src/main.rs:199:9
      <opt_dist::timer::TimerSection>::section::<opt_dist::execute_pipeline::{closure#1}, opt_dist::training::RustcPGOProfile>
             at /rustc/3d74566bec1c4a843859581f252317eb7b1bacbf/src/tools/opt-dist/src/timer.rs:111:22
      opt_dist::execute_pipeline
             at /rustc/3d74566bec1c4a843859581f252317eb7b1bacbf/src/tools/opt-dist/src/main.rs:196:29
             at /rustc/3d74566bec1c4a843859581f252317eb7b1bacbf/src/tools/opt-dist/src/main.rs:385:18
   4: <fn() -> core::result::Result<(), anyhow::Error> as core::ops::function::FnOnce<()>>::call_once
             at /rustc/782883f609713fe9617ba64d90086742ec62d374/library/core/src/ops/function.rs:250:5
             at /rustc/782883f609713fe9617ba64d90086742ec62d374/library/core/src/ops/function.rs:250:5
      std::sys_common::backtrace::__rust_begin_short_backtrace::<fn() -> core::result::Result<(), anyhow::Error>, core::result::Result<(), anyhow::Error>>
   5: std::rt::lang_start::<core::result::Result<(), anyhow::Error>>::{closure#0}
             at /rustc/782883f609713fe9617ba64d90086742ec62d374/library/std/src/rt.rs:167:18
             at /rustc/782883f609713fe9617ba64d90086742ec62d374/library/std/src/rt.rs:167:18
   6: core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &F>::call_once
      std::panicking::try::do_call
             at /rustc/782883f609713fe9617ba64d90086742ec62d374/library/std/src/panicking.rs:552:40
      std::panicking::try
             at /rustc/782883f609713fe9617ba64d90086742ec62d374/library/std/src/panicking.rs:516:19
```

</details>

aliemjay commented:
> The job **`x86_64-gnu-llvm-16`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/7092326727/job/19303382247) [(plain)](https://github.com/rust-lang/rust/commit/bcbfbe62ff1ca24302e585679473c43f59054717/checks/19303382247/logs)

MCVE:
```rust
struct VarZeroVec<'a, F>(&'a (), F);

trait Yokeable<'a> {
    type Output;
}

impl<'a> Yokeable<'a> for () {
    type Output = VarZeroVec<'a, ()>;
}

fn transform_mut<'a, F>(f: F)
where
    F: FnOnce(VarZeroVec<'a, ()>),
{
    f(absurd::<<() as Yokeable<'a>>::Output>())
}

fn absurd<T>() -> T{ todo!() }
```

aliemjay committed:
don't resolve regions in query input

fixes a soundness regression described in the PR description.

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit 8b85053888b0baee7da134fd652c433b1d287b79 with merge fd72347c63ef586b13bce5e2c07d442a0b5ec7ef...
<!-- homu: {"type":"TryBuildStarted","head_sha":"8b85053888b0baee7da134fd652c433b1d287b79","merge_sha":"fd72347c63ef586b13bce5e2c07d442a0b5ec7ef"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/7102836142/job/19337625292)
Build commit: fd72347c63ef586b13bce5e2c07d442a0b5ec7ef (`fd72347c63ef586b13bce5e2c07d442a0b5ec7ef`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/7102836142/job/19337625292"},"merge_sha":"fd72347c63ef586b13bce5e2c07d442a0b5ec7ef"} -->

rust-timer commented:
Queued fd72347c63ef586b13bce5e2c07d442a0b5ec7ef with parent 154f645d76a9752bd7ea58d7382b2e90e83557fb, future [comparison URL](https://perf.rust-lang.org/compare.html?start=154f645d76a9752bd7ea58d7382b2e90e83557fb&end=fd72347c63ef586b13bce5e2c07d442a0b5ec7ef).
There is currently 1 preceding artifact in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~1.9 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (fd72347c63ef586b13bce5e2c07d442a0b5ec7ef): [comparison URL](https://perf.rust-lang.org/compare.html?start=154f645d76a9752bd7ea58d7382b2e90e83557fb&end=fd72347c63ef586b13bce5e2c07d442a0b5ec7ef&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - ACTION NEEDED


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=154f645d76a9752bd7ea58d7382b2e90e83557fb&end=fd72347c63ef586b13bce5e2c07d442a0b5ec7ef&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.3%  | [0.3%, 0.4%]   | 9     |
| Regressions ❌ <br /> (secondary)  | 0.4%  | [0.2%, 1.1%]   | 16    |
| Improvements ✅ <br /> (primary)   | -2.3% | [-4.7%, -0.4%] | 24    |
| Improvements ✅ <br /> (secondary) | -0.8% | [-1.0%, -0.3%] | 3     |
| All ❌✅ (primary)                 | -1.6% | [-4.7%, 0.4%]  | 33    |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=154f645d76a9752bd7ea58d7382b2e90e83557fb&end=fd72347c63ef586b13bce5e2c07d442a0b5ec7ef&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 2.1%  | [2.1%, 2.1%]   | 1     |
| Regressions ❌ <br /> (secondary)  | 2.3%  | [1.6%, 3.0%]   | 2     |
| Improvements ✅ <br /> (primary)   | -0.9% | [-1.0%, -0.9%] | 2     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | 0.1%  | [-1.0%, 2.1%]  | 3     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=154f645d76a9752bd7ea58d7382b2e90e83557fb&end=fd72347c63ef586b13bce5e2c07d442a0b5ec7ef&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 0.8%  | [0.8%, 0.8%]   | 1     |
| Improvements ✅ <br /> (primary)   | -1.8% | [-2.6%, -1.3%] | 14    |
| Improvements ✅ <br /> (secondary) | -0.6% | [-0.6%, -0.6%] | 1     |
| All ❌✅ (primary)                 | -1.8% | [-2.6%, -1.3%] | 14    |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=154f645d76a9752bd7ea58d7382b2e90e83557fb&end=fd72347c63ef586b13bce5e2c07d442a0b5ec7ef&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 674.255s -> 673.753s (-0.07%)
**Artifact size**: 314.15 MiB -> 314.21 MiB (0.02%)

aliemjay committed:
fix small perf regressions

aliemjay committed:
don't store OriginalQueryValues::universe_map

ParamEnv is canonicalized in *queries input* rather than query response.
In such case we don't "preserve universes" of canonical variable.
This means that `universe_map` always has the default value, which is
wasteful to store in the cache.

aliemjay commented:
changed to fix the small regressions
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit 755b9dbd4d95c4d28e652ec2a1b070a06aed0783 with merge c68e68d3702abfa3bebacf9edf7c7fd27f9b3c3a...
<!-- homu: {"type":"TryBuildStarted","head_sha":"755b9dbd4d95c4d28e652ec2a1b070a06aed0783","merge_sha":"c68e68d3702abfa3bebacf9edf7c7fd27f9b3c3a"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/7113301969/job/19367646467)
Build commit: c68e68d3702abfa3bebacf9edf7c7fd27f9b3c3a (`c68e68d3702abfa3bebacf9edf7c7fd27f9b3c3a`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/7113301969/job/19367646467"},"merge_sha":"c68e68d3702abfa3bebacf9edf7c7fd27f9b3c3a"} -->

rust-timer commented:
Queued c68e68d3702abfa3bebacf9edf7c7fd27f9b3c3a with parent 17a520aa3769ebbf6653eef8369ab3c459b887aa, future [comparison URL](https://perf.rust-lang.org/compare.html?start=17a520aa3769ebbf6653eef8369ab3c459b887aa&end=c68e68d3702abfa3bebacf9edf7c7fd27f9b3c3a).
There is currently 1 preceding artifact in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~1.7 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (c68e68d3702abfa3bebacf9edf7c7fd27f9b3c3a): [comparison URL](https://perf.rust-lang.org/compare.html?start=17a520aa3769ebbf6653eef8369ab3c459b887aa&end=c68e68d3702abfa3bebacf9edf7c7fd27f9b3c3a&stat=instructions:u).

### Overall result: ✅ improvements - no action needed


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

@bors rollup=never
@rustbot label: -S-waiting-on-perf -perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=17a520aa3769ebbf6653eef8369ab3c459b887aa&end=c68e68d3702abfa3bebacf9edf7c7fd27f9b3c3a&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -1.9% | [-4.7%, -0.2%] | 31    |
| Improvements ✅ <br /> (secondary) | -1.5% | [-1.5%, -1.4%] | 2     |
| All ❌✅ (primary)                 | -1.9% | [-4.7%, -0.2%] | 31    |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=17a520aa3769ebbf6653eef8369ab3c459b887aa&end=c68e68d3702abfa3bebacf9edf7c7fd27f9b3c3a&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 4.0%  | [4.0%, 4.0%]   | 1     |
| Improvements ✅ <br /> (primary)   | -0.9% | [-0.9%, -0.9%] | 2     |
| Improvements ✅ <br /> (secondary) | -2.5% | [-2.5%, -2.5%] | 1     |
| All ❌✅ (primary)                 | -0.9% | [-0.9%, -0.9%] | 2     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=17a520aa3769ebbf6653eef8369ab3c459b887aa&end=c68e68d3702abfa3bebacf9edf7c7fd27f9b3c3a&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -1.8% | [-2.9%, -1.2%] | 14    |
| Improvements ✅ <br /> (secondary) | -2.0% | [-2.3%, -1.7%] | 2     |
| All ❌✅ (primary)                 | -1.8% | [-2.9%, -1.2%] | 14    |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=17a520aa3769ebbf6653eef8369ab3c459b887aa&end=c68e68d3702abfa3bebacf9edf7c7fd27f9b3c3a&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 674.58s -> 675.695s (0.17%)
**Artifact size**: 314.17 MiB -> 314.23 MiB (0.02%)

aliemjay committed:
remove canonicalize_query_preserving_universes

unused!

aliemjay committed:
make infcx optional in canonicalizer

This doesn't change behavior.
It should prevent unintentional resolution of inference variables
during canonicalization, which previously caused a soundness bug.
See PR description for more.

lcnr commented on compiler/rustc_infer/src/infer/canonical/canonicalizer.rs:
```diff
@@ -126,19 +126,52 @@ impl<'tcx> InferCtxt<'tcx> {
+                    param_env,
+                    self,
+                    self.tcx,
+                    &CanonicalizeFreeRegionsOtherThanStatic,
```
This does not spark joy 😁 Does anything break if you change this to `CanonicalizeFreeRegions`? I want to get away from canonicalizing static completely. iirc the last time I tried after #102472 there was still some failure in project or sth 🤔 would prefer for you (or me) to first try to  also canonicalize `'static` again to check why it matters.

aliemjay replied:
Borrowck error in [`tests/ui/trivial-bounds/trivial-bounds-object.rs`](https://github.com/rust-lang/rust/blob/5f73b006d2ebdb9fb245d952a40abac4a68ba3ec/tests/ui/trivial-bounds/trivial-bounds-object.rs)

ICE in [`tests/ui/nll/issue-61311-normalize.rs`](https://github.com/rust-lang/rust/blob/5f73b006d2ebdb9fb245d952a40abac4a68ba3ec/tests/ui/nll/issue-61311-normalize.rs) and [`tests/ui/nll/issue-61320-normalize.rs`](https://github.com/rust-lang/rust/blob/5f73b006d2ebdb9fb245d952a40abac4a68ba3ec/tests/ui/nll/issue-61320-normalize.rs). MCVE:
```rust
pub trait HasProj {
    type Proj;
}

impl<T: ?Sized> HasProj for T {
    type Proj = ();
}

fn test()
where
    dyn Send + 'static: HasProj,
    <dyn Send + 'static as HasProj>::Proj: Clone,
{
    let _ = test;
}
```

aliemjay replied:
cc #118965. It is also the root cause behind the ICE.

lcnr commented on compiler/rustc_infer/src/infer/canonical/canonicalizer.rs:
```diff
@@ -848,14 +846,8 @@ impl<'cx, 'tcx> Canonicalizer<'cx, 'tcx> {
-            let var = self.canonical_var(info, ty_var.into());
-            Ty::new_bound(self.tcx, self.binder_index, var.into())
-        }
+        let var = self.canonical_var(info, ty_var.into());
```
can you debug assert that `ty_var == infcx.shallow_resolve(ty_var)`?

lcnr commented on compiler/rustc_infer/src/infer/canonical/canonicalizer.rs:
```diff
@@ -867,13 +859,7 @@ impl<'cx, 'tcx> Canonicalizer<'cx, 'tcx> {
-            let var = self.canonical_var(info, const_var.into());
-            ty::Const::new_bound(self.tcx, self.binder_index, var, self.fold_ty(const_var.ty()))
-        }
+        let var = self.canonical_var(info, const_var.into());
```
same here

aliemjay replied:
done

lcnr reviewed (commented):
some nits, then r=me, afaict CI was an unrelated issue

aliemjay committed:
rustdoc: avoid ParamEnv with infer vars

ParamEnv's with inference variabels are invalid.

aliemjay commented:
> some nits, then r=me, afaict CI was an unrelated issue

I turns out it is related. Rustdoc incorrectly uses ParamEnv's with inference variables in them. This causes an ICE during canonicalization in one of the `unwrap`s in the commit "make infcx optional in canonicalizer". This was fixed by the last commit by using an empty param env. I assume it does not need a re-approval.

@bors r=lcnr rollup=never

bors commented:
:pushpin: Commit aa36c35296561db171939a6dc95279214fca3775 has been approved by `lcnr`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=lcnr aa36c35296561db171939a6dc95279214fca3775 -->
<!-- homu: {"type":"Approved","sha":"aa36c35296561db171939a6dc95279214fca3775","approver":"lcnr","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit aa36c35296561db171939a6dc95279214fca3775 with merge d23e1a689426638a5146c204e09278433c509610...
<!-- homu: {"type":"BuildStarted","head_sha":"aa36c35296561db171939a6dc95279214fca3775","merge_sha":"d23e1a689426638a5146c204e09278433c509610"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/7204323972/job/19627527846)
Approved by: lcnr
Pushing d23e1a689426638a5146c204e09278433c509610 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"lcnr","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/7204323972/job/19627527846"},"merge_sha":"d23e1a689426638a5146c204e09278433c509610"} -->

rust-timer commented:
Finished benchmarking commit (d23e1a689426638a5146c204e09278433c509610): [comparison URL](https://perf.rust-lang.org/compare.html?start=e6d1b0ec9859e6f5c29aaa3b6525fb625bf354ad&end=d23e1a689426638a5146c204e09278433c509610&stat=instructions:u).

### Overall result: ✅ improvements - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=e6d1b0ec9859e6f5c29aaa3b6525fb625bf354ad&end=d23e1a689426638a5146c204e09278433c509610&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -2.0% | [-4.8%, -0.2%] | 32    |
| Improvements ✅ <br /> (secondary) | -1.3% | [-1.4%, -1.3%] | 2     |
| All ❌✅ (primary)                 | -2.0% | [-4.8%, -0.2%] | 32    |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=e6d1b0ec9859e6f5c29aaa3b6525fb625bf354ad&end=d23e1a689426638a5146c204e09278433c509610&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.9% | [0.7%, 1.2%] | 2     |
| Regressions ❌ <br /> (secondary)  | 3.5% | [2.5%, 4.4%] | 2     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 0.9% | [0.7%, 1.2%] | 2     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=e6d1b0ec9859e6f5c29aaa3b6525fb625bf354ad&end=d23e1a689426638a5146c204e09278433c509610&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -1.8% | [-3.0%, -0.7%] | 16    |
| Improvements ✅ <br /> (secondary) | -1.5% | [-1.5%, -1.5%] | 1     |
| All ❌✅ (primary)                 | -1.8% | [-3.0%, -0.7%] | 16    |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=e6d1b0ec9859e6f5c29aaa3b6525fb625bf354ad&end=d23e1a689426638a5146c204e09278433c509610&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 673.665s -> 670.811s (-0.42%)
**Artifact size**: 312.46 MiB -> 312.42 MiB (-0.01%)
